controller/src/main/java/io/pravega/controller/store/stream/PravegaTablesStream.java

```diff
@@ -0,0 +1,811 @@
+                                               }));
+        }
+    }
+
+    private CompletableFuture<String> getMetadataTable() {
+        return getId().thenApply(this::getMetadataTable);
+    }
+
+    private String getMetadataTable(String id) {
+        return String.format(METADATA_TABLE, getName(), id);
+    }
+
+    private CompletableFuture<String> getEpochsWithTransactionsTable() {
+        return getId().thenApply(this::getEpochsWithTransactionsTable);
+    }
+
+    private String getEpochsWithTransactionsTable(String id) {
+        return String.format(EPOCHS_WITH_TRANSACTIONS_TABLE, getName(), id);
+    }
+
+    private CompletableFuture<String> getEpochTransactionsTable(int epoch) {
+
+        return getId().thenApply(id -> getEpochTransactionsTable(epoch, id));
+    }
+
+    private String getEpochTransactionsTable(int epoch, String id) {
+        return String.format(EPOCH_TRANSACTIONS_TABLE_FORMAT, getName(), id, epoch);
+    }
+
+    // region overrides
+
+    @Override
+    public CompletableFuture<Integer> getNumberOfOngoingTransactions() {
+        List<CompletableFuture<Integer>> futures = new ArrayList<>();
+        return getEpochsWithTransactionsTable()
+                .thenCompose(table -> storeHelper.getAllKeys(getScope(), table)
+                                              .collectRemaining(x -> {
+                                                  futures.add(getNumberOfOngoingTransactions(Integer.parseInt(x)));
+                                                  return true;
+                                              })
+                                              .thenCompose(v -> Futures.allOfWithResults(futures)
+                                                                       .thenApply(list -> list.stream().reduce(0, Integer::sum))));
+    }
+    
+    private CompletableFuture<Integer> getNumberOfOngoingTransactions(int epoch) {
+        AtomicInteger count = new AtomicInteger(0);
+        return getEpochTransactionsTable(epoch)
+                .thenCompose(tableName -> storeHelper.getAllKeys(getScope(), tableName).collectRemaining(x -> {
+                    count.incrementAndGet();
+                    return true;
+                }).thenApply(x -> count.get()));
+    }
+
+    @Override
+    public CompletableFuture<Void> deleteStream() {
+        // delete all tables for this stream even if they are not empty!
+        // 1. read epoch table 
+        // delete all epoch txn specific tables
+        // delete epoch txn base table
+        // delete metadata table
+        
+        // delete stream in scope 
+        String scope = getScope();
+        List<CompletableFuture<Void>> futures = new ArrayList<>();
+        return getId()
+                .thenCompose(id -> {
+                    String epochWithTxnTable = getEpochsWithTransactionsTable(id);
+                    return Futures.exceptionallyExpecting(storeHelper.getAllKeys(getScope(), epochWithTxnTable)
```

Comment:
This sequence does look restartable, but I just want to check with you on that.

Reply:
yes